### PR TITLE
Restrict print templates

### PR DESCRIPTION
**Pull request by NickGrantONS**
# Motivation and Context
Allowing rules to be set up which are wrong, or fulfilments to be requested which are wrong, is a huge problem for a system which is supposed to be able to be used by non-technical business users, and require minimal support. We need to restrict the UI to only allow things which are correct, and as such, we need to make the correct choices available in the UI to the users.

# What has changed
Added concept of pack codes which are 'allowed' for a survey... either as a scheduled action rule, or as a fulfilment.

# How to test?
Define a pack code like this:
```
insert into casev3.print_template
values ('TEST_PACK', '["__uac__"]', 'SUPPLIER_A');
```

Then, allow it to be used by fulfilments like this:
```
insert into casev3.fulfilment_survey_print_template
values ('<unique UUID>', 'TEST_PACK', '<survey ID>');
```

Then go ahead and request fulfilments for TEST_PACK pack code.

Or, allow the pack code to be used in print action rules like this:
```
insert into casev3.action_rule_survey_print_template
values ('<unique UUID>', 'TEST_PACK', '<survey ID>');
```

# Links
Trello: https://trello.com/c/o2vCZafV

### CODE REVIEW

**Comment by NickGrantONS** [comment]
[Process fulfilment's revenge: Process Fulfilment 2](https://trello.com/c/o2vCZafV/2494-process-fulfilments-revenge-process-fulfilment-2)

**Review by AdamHawtin — COMMENTED** [review]
Could this PR be pointed at the branch it is based on, until that branch is merged? It quite difficult to review when another branches changes are included in the diff.

**Review by AdamHawtin — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/ons/ssdc/caseprocessor/schedule/ActionRuleProcessor.java [hunk 1]
```diff
@@ -17,9 +17,8 @@ public ActionRuleProcessor(
     this.actionRuleRepository = actionRuleRepository;
   }
 
-  @Transactional(
-      propagation = Propagation.REQUIRES_NEW) // Start a new transaction for every action rule
-  public void processTriggeredActionRule(ActionRule triggeredActionRule) {
+  @Transactional(propagation = Propagation.REQUIRES_NEW) // Start a new transaction for every WoC
+  public void createScheduledActionRule(ActionRule triggeredActionRule) {
```
[the remark this answers is not in the document]
Comment (NickGrantONS) [reply src/main/java/uk/gov/ons/ssdc/caseprocessor/schedule/ActionRuleProcessor.java]: sorted

**Review by hbrace — APPROVED** [review]

File: src/main/java/uk/gov/ons/ssdc/caseprocessor/schedule/FulfilmentProcessor.java [hunk 2]
```diff
@@ -23,22 +23,22 @@ public FulfilmentProcessor(
...
+    List<String> packCodes = fulfilmentToProcessRepository.findDistinctPackCode();
 
-    fulfilmentCodes.forEach(
-        fulfilmentCode -> {
+    packCodes.forEach(
+        packCode -> {
           UUID batchId = UUID.randomUUID();
           log.with("batch_id", batchId)
-              .with("fulfilment_code", fulfilmentCode)
+              .with("fulfilment_code", packCode)
```
Comment (LukeLoze1) [inline src/main/java/uk/gov/ons/ssdc/caseprocessor/schedule/FulfilmentProcessor.java:32]: Should this be  .with("pack code", packCode)?

**Review by LukeLoze1 — COMMENTED** [review]
Minor question around logging

File: src/main/java/uk/gov/ons/ssdc/caseprocessor/schedule/FulfilmentProcessor.java (hunk 2 above)
Comment (NickGrantONS) [reply src/main/java/uk/gov/ons/ssdc/caseprocessor/schedule/FulfilmentProcessor.java]: fixed

**Review by LukeLoze1 — APPROVED** [review]

**Review by AdamHawtin — APPROVED** [review]

**Pull request merged**